File: frontend/src/services/assetBorrowService.js
import axiosClient from '~/utils/axiosClient';

const API_URL = '/asset-borrows';

class AssetBorrowService {
    async createAssetBorrow(assetId, request) {
        try {
            const res = await axiosClient.post(`${API_URL}?assetId=${assetId}`, request);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi tạo đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async updateAssetBorrow(id, request) {
        try {
            const res = await axiosClient.put(`${API_URL}/${id}`, request);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi cập nhật đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async getAllAssetBorrows(params = {}) {
        try {
            const res = await axiosClient.get(`${API_URL}`, { params });
            return res.data.data;
        } catch (error) {
            const errorMessage =
                error?.response?.data?.message || error.message || 'Lỗi khi lấy danh sách đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async getAssetBorrowById(id) {
        try {
            const res = await axiosClient.get(`${API_URL}/${id}`);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi lấy đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async getAssetBorrowsByRequestedById(requestedById) {
        try {
            const res = await axiosClient.get(`${API_URL}/requested-by/${requestedById}`);
            return res.data.data;
        } catch (error) {
            const errorMessage =
                error?.response?.data?.message || error.message || 'Lỗi khi lấy danh sách đơn mượn theo nhân viên';
            throw new Error(errorMessage);
        }
    }

    async getAssetBorrowsByAssetId(assetId) {
        try {
            const res = await axiosClient.get(`${API_URL}/asset/${assetId}`);
            return res.data.data;
        } catch (error) {
            const errorMessage =
                error?.response?.data?.message || error.message || 'Lỗi khi lấy danh sách đơn mượn theo tài sản';
            throw new Error(errorMessage);
        }
    }

    async approveAssetBorrow(id) {
        try {
            const res = await axiosClient.put(`${API_URL}/${id}/approve`);
            return res.data.data;
        } catch (error) {
            const errorMessage =
                error?.response?.data?.message || error.message || 'Lỗi khi phê duyệt đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async rejectAssetBorrow(id, data) {
        try {
            const res = await axiosClient.put(`${API_URL}/${id}/reject`, data);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi từ chối đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }

    async returnAsset(id) {
        try {
            const res = await axiosClient.put(`${API_URL}/${id}/return`);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi trả tài sản';
            throw new Error(errorMessage);
        }
    }

    async deleteAssetBorrow(id) {
        try {
            const res = await axiosClient.delete(`${API_URL}/${id}`);
            return res.data.data;
        } catch (error) {
            const errorMessage = error?.response?.data?.message || error.message || 'Lỗi khi xóa đơn mượn tài sản';
            throw new Error(errorMessage);
        }
    }
}

const assetBorrowService = new AssetBorrowService();
export default assetBorrowService;

File: backend/src/main/java/htmp/codien/quanlycodien/modules/asset/service/AssetBorrowService.java
package htmp.codien.quanlycodien.modules.asset.service;

import htmp.codien.quanlycodien.modules.asset.dto.borrow.AssetBorrowCreationRequest;
import htmp.codien.quanlycodien.modules.asset.dto.borrow.AssetBorrowRejectRequest;
import htmp.codien.quanlycodien.modules.asset.dto.borrow.AssetBorrowResponse;
import htmp.codien.quanlycodien.modules.asset.dto.borrow.AssetBorrowUpdationRequest;
import htmp.codien.quanlycodien.modules.asset.entity.AssetBorrow;
import htmp.codien.quanlycodien.modules.asset.enums.AssetBorrowStatus;

import java.time.LocalDate;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface AssetBorrowService {
    void createAssetBorrow(Long assetId, AssetBorrowCreationRequest request);

    void updateAssetBorrow(Long id, AssetBorrowUpdationRequest request);

    Page<AssetBorrowResponse> getAllAssetBorrows(Pageable pageable,
            String keyword,
            Long requestedById,
            LocalDate date,
            LocalDate borrowDate,
            AssetBorrowStatus status);

    AssetBorrowResponse getAssetBorrowById(Long id);

    void deleteAssetBorrow(Long id);

    List<AssetBorrowResponse> getAssetBorrowsByRequestedById(Long requestedById);

    List<AssetBorrowResponse> getAssetBorrowsByAssetId(Long assetId);

    void updateStatusAssetBorrow(Long id, AssetBorrowStatus status);

    void returnAsset(Long id);

    AssetBorrow findEntityById(Long id);

    void approveAssetBorrow(Long id);

    void rejectAssetBorrow(Long id, AssetBorrowRejectRequest req);
}